### DISCUSSION TITLE
Force flush traces from agent TracerProvider

### DISCUSSION THREAD

**Discussion by PaddyDrury** [open-telemetry/opentelemetry-java-instrumentation#7865] [Q&A] [answered]
Is there a mechanism for force flushing traces from the java agent tracer provider?

I have cucumber acceptance tests which are auto-instrumented using the java agent. The application under test is also instrumented and the spans for both are sent to the same jaeger instance.

For each cucumber scenario, using some manual instrumentation, I start a trace and have a span for each step. After completion of a scenario I end the scenario trace.

What I would like to be able to do after ending the scenario trace is to force flush all related spans, and then call the jaeger traces API to add the trace json as an attachment to the scenario for the cucumber test report.

The problem I'm encountering is that the agent `TracerProvider` returned by `GlobalOpenTelemetry.getTracerProvider()` does not implement `forceFlush()` as per the `SdkTracerProvider`. When I debug the tests it looks like nested within the agent `TracerProvider` there is a delegate `SdkTracerProvider` but I cannot find a way of accessing it. Any help on this would be appreciated.

**Comment by laurit** [comment, marked as answer]
See https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/4d202ba7e9a182b8c8de2d4aba0c60722a785be5/instrumentation/aws-lambda/aws-lambda-core-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/awslambdacore/v1_0/AwsLambdaRequestHandlerInstrumentation.java#L85 Add compile time dependency to https://search.maven.org/artifact/io.opentelemetry.javaagent/opentelemetry-javaagent-bootstrap/1.23.0-alpha/jar to use `OpenTelemetrySdkAccess` or use reflection to call it if you don't want to add dependency to `opentelemetry-javaagent-bootstrap`. If you wish to run the same test without agent use `Class.forName` to test whether `OpenTelemetrySdkAccess` or some other agent class is present before calling `OpenTelemetrySdkAccess`.

**Comment by PaddyDrury** [comment]
Thank you. That's what I'm after.
